httpx/middleware/auth.py

```diff
@@ -0,0 +1,219 @@
+import hashlib
+import os
+import time
+import typing
+from base64 import b64encode
+from urllib.request import parse_http_list
+
+from ..exceptions import ProtocolError
+from ..models import AsyncRequest, AsyncResponse, StatusCode
+from ..utils import to_bytes, to_str, unquote
+from .base import BaseMiddleware
+
+
+class BasicAuthMiddleware(BaseMiddleware):
+    def __init__(
+        self, username: typing.Union[str, bytes], password: typing.Union[str, bytes]
+    ):
+        if isinstance(username, str):
+            username = username.encode("latin1")
```

Comment:
We should probably use UTF-8 here given this comment: https://github.com/encode/httpx/pull/259#discussion_r320645771

Eventually we should merge the two implementations, I can do that as a part of my PR once this lands. :)